### PR TITLE
[ios|android][ads-admob] added Server Side Verification Options

### PR DESCRIPTION
**Pull request by swittk**
# Why

Admob rewarded ads, when used online, require server side verification to work properly

# How

- Added parameters to requestAdAsync
- Added additional argument to native requestAd

# Test Plan
- Planned to test on simulator, unfortunately I'm stuck on XCode 11 (Mojave) for the time being and can't compile the current BareExpo project due to multiple headers not being available. The changes are minimal and should compile and run as expected in the right environment though.
- Android : Can compile.

# Checklist
- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

** Note : Please do test this again, I am currently limited in testing this implementation due to Xcode 12 limitations, and lack of time. The code should work on iOS, but for Android I'll need someone to check if the checks on the ReadableArguments object are the correct way to do so (I can't find any docs on ReadableArguments, so I had to imply from the original class)

### CODE REVIEW

**Comment by swittk** [comment]
Um.. hi, I've tried the build in Catalina now, and it builds. And runs.

However I've found this error which appears either in the clean master branch or my own branch. Effectively, the Admob module isn't loaded at all and attempting to access anything in the module would say that it is null. This seems to be an issue that has popped up before but I have no idea why it's happening now https://github.com/expo/expo/issues/6860

Screen Shot 2564-04-26 at 20 47 25

Simulator Screen Shot - iPhone 11 - 2021-04-26 at 20 37 14

@tsapeta  Would you mind taking a look?  I simply can't test it due to this issue with unimodules. These code changes are very minor but would benefit all users of AdMob that use Server Side Verification. I can assure you that the code builds without errors.

**Comment by cruzach** [comment]
@swittk can you link the documentation for server side request verification for admob native libraries?

**Comment by swittk** [comment]
Sure !

Admob Server-side verification works by the AdMob server making a call to our servers once the Admob has determined that the client has finished viewing the ad.  ([Documentation](https://support.google.com/admob/answer/9603226?hl=en))  
- Example call :
````
https://www.myserver.com/path?ad_network=54...55&ad_unit=12345678&reward_amount=10&reward_item=coins
&timestamp=150777823&transaction_id=12...DEF&user_id=1234567&signature=ME...Z1c&key_id=1268887
````
However, the callback would by default only call with no parameters identifying which user/device (specified as the user_id part of the query string) to reward, or any extra string data (the custom_data query string part). To make google call our server with these parameters in the query string, we must pass them to the initial requests to the Mobile Ads SDK in the [GADRewardedAd (iOS)](https://developers.google.com/admob/ios/rewarded-ads) class, or the [RewardedAd (Android)](https://developers.google.com/ad-manager/mobile-ads-sdk/android/rewarded-fullscreen#always_test_with_test_ads) class.

- iOS
  - Appending a custom_data payload : https://developers.google.com/admob/ios/rewarded-video-ssv#custom_data
  - user_id is set similarly, as seen in this list here : https://developers.google.com/admob/ios/rewarded-video-ssv#ssv_callback_parameters

- Android
  - Appending a custom_data payload : https://developers.google.com/admob/android/rewarded-video-ssv#custom_data
  - user_id setting (similar to iOS) : https://developers.google.com/admob/android/rewarded-video-ssv#ssv_callback_parameters

File: packages/expo-ads-admob/android/src/main/java/expo/modules/ads/admob/AdMobRewardedVideoAdModule.java [hunk 1]
```diff
@@ -112,6 +113,21 @@ public void run() {
 
         mRewardedAd = new RewardedAd(mActivityProvider.getCurrentActivity(), mAdUnitID);
 
+        if( ssvOptions != null && !ssvOptions.isEmpty() ) {
+          // Check if ReadableArguments has a value (see doc on ReadableArguments here https://github.com/expo/expo/blob/59794c1cc56ac8ea110c208ba0711dbd1e044594/packages/%40unimodules/core/android/src/main/java/org/unimodules/core/arguments/ReadableArguments.java#L10)
+          ServerSideVerificationOptions options = new ServerSideVerificationOptions.Builder();
+          String userId = ssvOptions.getString("userId");
+          String customData = ssvOptions.getString("customData");
+          if(userId != null) {
```
Comment (cruzach) [inline packages/expo-ads-admob/android/src/main/java/expo/modules/ads/admob/AdMobRewardedVideoAdModule.java:121]: what happens if we `setServerSideVerificationOptions` and either the `userId` or `customData` fields are missing?

File: packages/expo-ads-admob/ios/EXAdsAdMob/EXAdsAdMobRewarded.m [hunk 2]
```diff
@@ -85,6 +86,16 @@ - (void)stopObserving {
...
       [request registerAdNetworkExtras:extras];
     }
+    if(ssvOptions) {
+      GADServerSideVerificationOptions *options = [[GADServerSideVerificationOptions alloc] init];
+      if(ssvOptions[@"userId"]) {
+        options.userIdentifier = ssvOptions[@"userId"];
+      }
+      if(ssvOptions[@"customData"]) {
+        options.customRewardString = ssvOptions[@"customData"];
+      }
```
Comment (cruzach) [inline packages/expo-ads-admob/ios/EXAdsAdMob/EXAdsAdMobRewarded.m:96]: do we need to check and ensure each of these values that were provided by the user are a String? (assume not all users are using TS)

File: packages/expo-ads-admob/ios/EXAdsAdMob/EXAdsAdMobRewarded.m [hunk 3]
```diff
@@ -85,6 +86,16 @@ - (void)stopObserving {
...
+      self.rewardedAd.serverSideVerificationOptions = options;
```
Comment (cruzach) [inline packages/expo-ads-admob/ios/EXAdsAdMob/EXAdsAdMobRewarded.m:97]: same question as on the android side- what happens if one or both of `userId` or `customData` is missing? According to https://support.google.com/admob/answer/9603226?hl=en they are optional so I would assume nothing happens

**Review by cruzach — COMMENTED** [review]
thanks for taking the time to add this in! Could you have Xcode re-indent the file you changed? Should be under `Editor > Structure > Re-Indent`

File: packages/expo-ads-admob/android/src/main/java/expo/modules/ads/admob/AdMobRewardedVideoAdModule.java (hunk 1 above)
Comment (swittk) [reply packages/expo-ads-admob/android/src/main/java/expo/modules/ads/admob/AdMobRewardedVideoAdModule.java:121]: I've verified, it just performs all calls normally like the ServerSideVerificationOptions are not provided.

File: packages/expo-ads-admob/ios/EXAdsAdMob/EXAdsAdMobRewarded.m (hunk 3 above)
Comment (swittk) [reply packages/expo-ads-admob/ios/EXAdsAdMob/EXAdsAdMobRewarded.m:97]: I've verified this too, it just performs all calls normally as if the GADServerSideVerificationOptions are not provided. (Server callback is called without user_id or custom_data in the querystring)

> Attached heroku log
Screen Shot 2564-04-29 at 09 15 48

File: packages/expo-ads-admob/ios/EXAdsAdMob/EXAdsAdMobRewarded.m (hunk 2 above)
Comment (swittk) [reply packages/expo-ads-admob/ios/EXAdsAdMob/EXAdsAdMobRewarded.m]: Ok, I've added the explicit checks for [... isKindOfClass:[NSString class]].

**Comment by swittk** [comment]
It seems expo-ci has deleted my comments :(

The ServerSideVerificationOptions object can be assigned with no properties, and the ad request calls and server side calls will simply proceed as if no server side verification options were set.

**Comment by swittk** [comment]
@cruzach Sorry, I know you're likely busy, but I'm wondering if there are further changes needed for this code.

**Comment by Simek** [comment]
Closing, since the `ads-admob` package has been removed from the repository, see #18168.

**Pull request closed without merging**